File: managefarmspro/managefarmspro/report/resource_utilization_breakdown/resource_utilization_breakdown.py
# Copyright (c) 2025, Khalandar Sihan and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
import json
from frappe import _
from frappe.utils import getdate, add_months, nowdate, flt, get_first_day, get_last_day, formatdate
from datetime import datetime

def execute(filters=None):
    if not filters:
        filters = {}
        
    # Set default filters if not specified
    if not filters.get("from_date"):
        filters["from_date"] = add_months(nowdate(), -12)  # Default to last 12 months
    if not filters.get("to_date"):
        filters["to_date"] = nowdate()
        
    columns = get_columns(filters)
    data = get_data(filters)
    
    # Prepare chart data
    chart = get_chart(data, filters)
    
    return columns, data, None, chart

def get_columns(filters):
    group_by = filters.get("group_by", "Month")
    
    columns = [
        {
            "fieldname": group_by.lower(),
            "label": _(group_by),
            "fieldtype": "Data",
            "width": 120
        },
        {
            "fieldname": "total_cost",
            "label": _("Total Cost"),
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "fieldname": "labor_cost",
            "label": _("Labor Cost"),
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "fieldname": "labor_percentage",
            "label": _("Labor %"),
            "fieldtype": "Percent",
            "width": 120
        },
        {
            "fieldname": "material_cost",
            "label": _("Material Cost"),
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "fieldname": "material_percentage",
            "label": _("Material %"),
            "fieldtype": "Percent",
            "width": 120
        },
        {
            "fieldname": "equipment_cost",
            "label": _("Equipment Cost"),
            "fieldtype": "Currency",
            "width": 150
        },
        {
            "fieldname": "equipment_percentage",
            "label": _("Equipment %"),
            "fieldtype": "Percent",
            "width": 120
        },
        {
            "fieldname": "supervision_cost",
            "label": _("Supervision Cost"),
            "fieldtype": "Currency",
            "width": 150
        },
        {
            "fieldname": "supervision_percentage",
            "label": _("Supervision %"),
            "fieldtype": "Percent",
            "width": 120
        },
        {
            "fieldname": "work_count",
            "label": _("# of Works"),
            "fieldtype": "Int",
            "width": 120
        }
    ]
    
    return columns

def get_data(filters):
    conditions = get_conditions(filters)
    group_by = filters.get("group_by", "Month")
    
    # First, get all works that match the conditions
    works = frappe.db.sql("""
        SELECT 
            name,
            work_name,
            plot,
            total_cost,
            work_date,
            COALESCE(
                supervision_charges, 
                (SELECT supervision_charges FROM `tabPlot` WHERE name = `tabWork`.plot), 
                0
            ) as supervision_charges
        FROM 
            `tabWork`
        WHERE
            docstatus = 1
            {conditions}
        ORDER BY 
            work_date
    """.format(conditions=conditions), filters, as_dict=1)
    
    # Define the grouping function based on the selected group_by option
    def get_group_key(work):
        if group_by == "Month":
            # Format: "Jan 2025"
            return getdate(work.work_date).strftime("%b %Y")
        elif group_by == "Quarter":
            # Format: "Q1 2025"
            date = getdate(work.work_date)
            quarter = ((date.month - 1) // 3) + 1
            return f"Q{quarter} {date.year}"
        elif group_by == "Year":
            # Format: "2025"
            return str(getdate(work.work_date).year)
        elif group_by == "Work Type":
            # Return the work type
            return work.work_name
        elif group_by == "Plot":
            # Return the plot
            return work.plot
        elif group_by == "Cluster":
            # Get the cluster for the plot
            plot_info = frappe.db.get_value("Plot", work.plot, "cluster", as_dict=1)
            return plot_info.cluster if plot_info else "Unknown"
        else:
            # Default to Month
            return getdate(work.work_date).strftime("%b %Y")
    
    # Group works based on the selected criteria
    grouped_data = {}
    
    for work in works:
        group_key = get_group_key(work)
        
        if group_key not in grouped_data:
            grouped_data[group_key] = {
                group_by.lower(): group_key,
                "total_cost": 0,
                "labor_cost": 0,
                "material_cost": 0,
                "equipment_cost": 0,
                "supervision_cost": 0,
                "work_count": 0,
                "sort_key": work.work_date if group_by in ["Month", "Quarter", "Year"] else group_key
            }
        
        # Get cost breakdown for this work
        labor_cost, material_cost, equipment_cost = get_cost_breakdown(work.name)
        
        # Calculate supervision cost
        try:
            supervision_percentage = float(work.supervision_charges or 0)
        except (ValueError, TypeError):
            supervision_percentage = 0
            
        supervision_cost = (work.total_cost * supervision_percentage) / 100
        supervision_cost = round(supervision_cost, 2)
        
        # Add to group totals
        grouped_data[group_key]["total_cost"] += work.total_cost
        grouped_data[group_key]["labor_cost"] += labor_cost
        grouped_data[group_key]["material_cost"] += material_cost
        grouped_data[group_key]["equipment_cost"] += equipment_cost
        grouped_data[group_key]["supervision_cost"] += supervision_cost
        grouped_data[group_key]["work_count"] += 1
    
    # Convert to list and calculate percentages
    result = []
    for key, data in grouped_data.items():
        # Calculate percentages for resource distribution
        total = data["total_cost"]
        if total > 0:
            data["labor_percentage"] = round((data["labor_cost"] / total) * 100, 2)
            data["material_percentage"] = round((data["material_cost"] / total) * 100, 2)
            data["equipment_percentage"] = round((data["equipment_cost"] / total) * 100, 2)
            data["supervision_percentage"] = round((data["supervision_cost"] / total) * 100, 2)
        else:
            data["labor_percentage"] = 0
            data["material_percentage"] = 0
            data["equipment_percentage"] = 0
            data["supervision_percentage"] = 0
        
        result.append(data)
    
    # Sort the result based on group_by
    if group_by in ["Month", "Quarter", "Year"]:
        # For time-based grouping, sort by date
        result.sort(key=lambda x: x["sort_key"])
    else:
        # For other groupings, sort by total_cost (descending)
        result.sort(key=lambda x: x["total_cost"], reverse=True)
    
    # Remove the sort key from the final output
    for entry in result:
        if "sort_key" in entry:
            del entry["sort_key"]
    
    return result

def get_cost_breakdown(work_name):
    """Get the cost breakdown for a specific work"""
    labor_cost = frappe.db.sql("""
        SELECT COALESCE(SUM(total_price), 0) as cost
        FROM `tabLabor Child`
        WHERE parent = %s
    """, work_name, as_dict=1)[0].cost
    
    material_cost = frappe.db.sql("""
        SELECT COALESCE(SUM(total_price), 0) as cost
        FROM `tabMaterial Child`
        WHERE parent = %s
    """, work_name, as_dict=1)[0].cost
    
    equipment_cost = frappe.db.sql("""
        SELECT COALESCE(SUM(total_price), 0) as cost
        FROM `tabEquipment Child`
        WHERE parent = %s
    """, work_name, as_dict=1)[0].cost
    
    return labor_cost, material_cost, equipment_cost

def get_chart(data, filters):
    """Generate chart data for the report"""
    if not data:
        return None
    
    chart_type = filters.get("chart_type", "Stacked").lower()
    if chart_type == "percentage":
        chart_type = "pie"
    
    # Get the group_by field
    group_by = filters.get("group_by", "Month").lower()
    
    labels = []
    labor_values = []
    material_values = []
    equipment_values = []
    supervision_values = []
    
    for entry in data:
        labels.append(entry[group_by])
        labor_values.append(entry["labor_cost"])
        material_values.append(entry["material_cost"])
        equipment_values.append(entry["equipment_cost"])
        supervision_values.append(entry["supervision_cost"])
    
    # If it's a pie chart, we want to show the overall distribution rather than per group
    if chart_type == "pie":
        # Calculate totals
        total_labor = sum(labor_values)
        total_material = sum(material_values)
        total_equipment = sum(equipment_values)
        total_supervision = sum(supervision_values)
        
        # Create a simple pie chart of overall resource distribution
        chart = {
            "data": {
                "labels": ["Labor", "Material", "Equipment", "Supervision"],
                "datasets": [{
                    "values": [total_labor, total_material, total_equipment, total_supervision]
                }]
            },
            "type": "pie",
            "colors": ["#5EC962", "#7B6CFF", "#FFA00A", "#00BCD4"],
            "height": 300
        }
    else:
        # For bar, line, or stacked chart
        datasets = []
        chart_colors = ["#5EC962", "#7B6CFF", "#FFA00A", "#00BCD4"]
        
        # Determine if we should use stacked bar
        is_stacked = (chart_type == "stacked")
        
        # Use the appropriate chart type
        actual_chart_type = "bar" if chart_type in ["bar", "stacked"] else chart_type
        
        # Create datasets based on chart type
        datasets = [
            {"name": "Labor", "values": labor_values},
            {"name": "Material", "values": material_values},
            {"name": "Equipment", "values": equipment_values},
            {"name": "Supervision", "values": supervision_values}
        ]
        
        chart = {
            "data": {
                "labels": labels,
                "datasets": datasets
            },
            "type": actual_chart_type,
            "colors": chart_colors,
            "height": 300
        }
        
        # Add stacked property if needed
        if is_stacked:
            chart["stacked"] = 1
    
    return chart

def get_conditions(filters):
    """Build the WHERE clause conditions based on filters"""
    conditions = []
    
    if filters.get("from_date") and filters.get("to_date"):
        conditions.append(" AND work_date BETWEEN %(from_date)s AND %(to_date)s")
    
    if filters.get("cluster"):
        # This will need a join with Plot to get the cluster
        conditions.append(" AND plot IN (SELECT name FROM `tabPlot` WHERE cluster = %(cluster)s)")
    
    if filters.get("plot"):
        conditions.append(" AND plot = %(plot)s")
        
    if filters.get("customer"):
        conditions.append(" AND customer = %(customer)s")
        
    if filters.get("work_name"):
        conditions.append(" AND work_name = %(work_name)s")
    
    return " ".join(conditions)

@frappe.whitelist()
def get_resource_summary(filters):
    """API endpoint to get resource summary data for dashboards"""
    if isinstance(filters, str):
        filters = json.loads(filters)
    
    data = get_data(filters)
    
    # Calculate totals across all groups
    total_cost = sum(d["total_cost"] for d in data)
    labor_cost = sum(d["labor_cost"] for d in data)
    material_cost = sum(d["material_cost"] for d in data)
    equipment_cost = sum(d["equipment_cost"] for d in data)
    supervision_cost = sum(d["supervision_cost"] for d in data)
    
    # Calculate percentages
    labor_percentage = round((labor_cost / total_cost * 100), 2) if total_cost else 0
    material_percentage = round((material_cost / total_cost * 100), 2) if total_cost else 0
    equipment_percentage = round((equipment_cost / total_cost * 100), 2) if total_cost else 0
    supervision_percentage = round((supervision_cost / total_cost * 100), 2) if total_cost else 0
    
    return {
        "total_cost": total_cost,
        "labor": {"cost": labor_cost, "percentage": labor_percentage},
        "material": {"cost": material_cost, "percentage": material_percentage},
        "equipment": {"cost": equipment_cost, "percentage": equipment_percentage},
        "supervision": {"cost": supervision_cost, "percentage": supervision_percentage}
    }

File: managefarmspro/managefarmspro/report/resource_utilization_breakdown/resource_utilization_breakdown.js
// Copyright (c) 2025, Khalandar Sihan and contributors
// For license information, please see license.txt

frappe.query_reports["Resource Utilization Breakdown"] = {
  filters: [
    {
      fieldname: "from_date",
      label: __("From Date"),
      fieldtype: "Date",
      default: frappe.datetime.add_months(frappe.datetime.get_today(), -12),
      reqd: 1,
    },
    {
      fieldname: "to_date",
      label: __("To Date"),
      fieldtype: "Date",
      default: frappe.datetime.get_today(),
      reqd: 1,
    },
    {
      fieldname: "cluster",
      label: __("Cluster"),
      fieldtype: "Link",
      options: "Cluster",
      on_change: function () {
        let cluster = frappe.query_report.get_filter_value("cluster");

        if (!cluster) {
          // If cluster is cleared, also clear plot and customer
          frappe.query_report.set_filter_value("plot", "");
          frappe.query_report.set_filter_value("customer", "");
        } else {
          // When cluster changes, clear dependent filters but don't refresh yet
          frappe.query_report.set_filter_value("plot", "");
          frappe.query_report.set_filter_value("customer", "");

          // Force refresh of the customer dropdown options to reflect the new cluster
          setTimeout(function () {
            // This trick forces the customer field to re-evaluate its get_query function
            frappe.query_report.set_filter_value("customer", "");
          }, 100);
        }

        // Always refresh after changing filters
        frappe.query_report.refresh();
      },
    },
    {
      fieldname: "plot",
      label: __("Plot"),
      fieldtype: "Link",
      options: "Plot",
      get_query: function () {
        let cluster = frappe.query_report.get_filter_value("cluster");
        if (cluster) {
          return {
            filters: {
              cluster: cluster,
            },
          };
        }
      },
      on_change: function () {
        let plot = frappe.query_report.get_filter_value("plot");

        if (!plot) {
          // If plot is cleared, also clear customer and refresh
          frappe.query_report.set_filter_value("customer", "");
          frappe.query_report.refresh();
        } else {
          // If a plot is selected, automatically fetch and set the customer
          frappe.call({
            method:
              "managefarmspro.managefarmspro.report.monthly_work_cost_trend.monthly_work_cost_trend.get_customer_for_plot",
            args: {
              doctype: "Customer",
              txt: "",
              searchfield: "name",
              start: 0,
              page_len: 1,
              filters: JSON.stringify({ plot: plot }),
            },
            callback: function (r) {
              // If customer data was returned, set it as the customer filter value
              if (r.message && r.message.length > 0) {
                frappe.query_report.set_filter_value(
                  "customer",
                  r.message[0][0]
                );
              } else {
                // Clear customer if none found for this plot
                frappe.query_report.set_filter_value("customer", "");
              }
              // Refresh the report after setting/clearing the customer
              frappe.query_report.refresh();
            },
          });
        }
      },
    },
    {
      fieldname: "customer",
      label: __("Customer"),
      fieldtype: "Link",
      options: "Customer",
      get_query: function () {
        let plot = frappe.query_report.get_filter_value("plot");
        let cluster = frappe.query_report.get_filter_value("cluster");

        if (plot) {
          // If plot is selected, filter customer based on that plot
          return {
            query:
              "managefarmspro.managefarmspro.report.monthly_work_cost_trend.monthly_work_cost_trend.get_customer_for_plot",
            filters: {
              plot: plot,
            },
          };
        } else if (cluster) {
          // If only cluster is selected (no plot), filter customers based on cluster
          return {
            query:
              "managefarmspro.managefarmspro.report.monthly_work_cost_trend.monthly_work_cost_trend.get_customers_for_cluster",
            filters: {
              cluster: cluster,
            },
          };
        }
      },
      on_change: function () {
        // Refresh the report when customer changes
        frappe.query_report.refresh();
      },
    },
    {
      fieldname: "work_name",
      label: __("Work Type"),
      fieldtype: "Link",
      options: "Work Item",
    },
    {
      fieldname: "group_by",
      label: __("Group By"),
      fieldtype: "Select",
      options: "Month\nQuarter\nYear\nWork Type\nPlot\nCluster",
      default: "Month",
    },
    {
      fieldname: "chart_type",
      label: __("Chart Type"),
      fieldtype: "Select",
      options: "Bar\nLine\nPie\nPercentage\nStacked",
      default: "Stacked",
    },
  ],

  formatter: function (value, row, column, data, default_formatter) {
    value = default_formatter(value, row, column, data);

    if (
      [
        "labor_percentage",
        "equipment_percentage",
        "material_percentage",
        "supervision_percentage",
      ].includes(column.fieldname)
    ) {
      // Add color based on percentage value
      const percentage = parseFloat(data[column.fieldname]);

      if (percentage > 50) {
        value = `<span style="color: red; font-weight: bold;">${value}</span>`;
      } else if (percentage > 30) {
        value = `<span style="color: orange;">${value}</span>`;
      }
    }

    return value;
  },

  onload: function (report) {
    // Add custom buttons for visualization options
    report.page.add_inner_button(__("View as Absolute Values"), function () {
      toggleChartView(report, "absolute");
    });

    report.page.add_inner_button(__("View as Percentages"), function () {
      toggleChartView(report, "percentage");
    });

    report.page.add_inner_button(__("Monthly Trend"), function () {
      showMonthlyTrend(report);
    });

    report.page.add_inner_button(__("Resource Summary"), function () {
      showResourceSummary(report);
    });

    // report.page.add_inner_button(__("Export Chart"), function () {
    //   exportChart(report);
    // });

    // Hide the Actions button that comes by default
    setTimeout(function () {
      // Try multiple selectors to catch the Actions button
      //   $(".actions-btn-group").hide();
      //   $('.btn-group:contains("Actions")').hide();
      //   $('button:contains("Actions")').hide();
      //   $('.page-actions button:contains("Actions")').hide();
      //   $('.standard-actions button:contains("Actions")').hide();
      //   $('.custom-actions button:contains("Actions")').hide();
      // If there's a div with class containing 'actions'
      //   $('div[class*="actions"]:contains("Actions")').hide();
      // Remove it from DOM if hiding doesn't work
      //   $('.page-actions button:contains("Actions")').remove();
    }, 1000);
  },
};

// Toggle between absolute values and percentages in the chart
function toggleChartView(report, viewType) {
  if (!report.data || !report.data.length) {
    frappe.msgprint(__("No data available to display"));
    return;
  }

  let data = report.data;
  let labels = [];
  let laborValues = [];
  let materialValues = [];
  let equipmentValues = [];
  let supervisionValues = [];
  let chartTitle = "";

  // Extract labels and values based on group_by field
  let groupBy = frappe.query_report.get_filter_value("group_by") || "Month";

  data.forEach((d) => {
    labels.push(d[groupBy.toLowerCase()] || d.period);

    if (viewType === "percentage") {
      laborValues.push(d.labor_percentage || 0);
      materialValues.push(d.material_percentage || 0);
      equipmentValues.push(d.equipment_percentage || 0);
      supervisionValues.push(d.supervision_percentage || 0);
      chartTitle = "Resource Utilization (Percentage Breakdown)";
    } else {
      laborValues.push(d.labor_cost || 0);
      materialValues.push(d.material_cost || 0);
      equipmentValues.push(d.equipment_cost || 0);
      supervisionValues.push(d.supervision_cost || 0);
      chartTitle = "Resource Utilization (Absolute Values)";
    }
  });

  // Get current chart type
  let chartType =
    frappe.query_report.get_filter_value("chart_type") || "Stacked";
  let chartTypeForRender = chartType.toLowerCase();

  if (chartType === "Percentage" || chartType === "Stacked") {
    chartTypeForRender = "bar";
  }

  // Define chart colors
  let colors = ["#5EC962", "#7B6CFF", "#FFA00A", "#00BCD4"];

  // Create datasets
  let datasets = [];

  if (chartType === "Stacked" || chartType === "Bar" || chartType === "Line") {
    datasets = [
      {
        name: "Labor",
        values: laborValues,
        chartType: chartType === "Line" ? "line" : "bar",
      },
      {
        name: "Material",
        values: materialValues,
        chartType: chartType === "Line" ? "line" : "bar",
      },
      {
        name: "Equipment",
        values: equipmentValues,
        chartType: chartType === "Line" ? "line" : "bar",
      },
      {
        name: "Supervision",
        values: supervisionValues,
        chartType: chartType === "Line" ? "line" : "bar",
      },
    ];
  } else {
    // For pie charts, calculate the total for each resource type
    const totalLabor = laborValues.reduce((sum, val) => sum + val, 0);
    const totalMaterial = materialValues.reduce((sum, val) => sum + val, 0);
    const totalEquipment = equipmentValues.reduce((sum, val) => sum + val, 0);
    const totalSupervision = supervisionValues.reduce(
      (sum, val) => sum + val,
      0
    );

    labels = ["Labor", "Material", "Equipment", "Supervision"];
    datasets = [
      {
        values: [totalLabor, totalMaterial, totalEquipment, totalSupervision],
      },
    ];
  }

  // Create chart data
  let chartData = {
    labels: labels,
    datasets: datasets,
  };

  // Chart options
  let chartOptions = {
    height: 300,
    colors: colors,
  };

  // Add special formatting for percentage view
  if (viewType === "percentage") {
    chartOptions.tooltipOptions = {
      formatTooltipY: (d) => d.toFixed(1) + "%",
    };
    chartOptions.axisOptions = {
      yAxisMode: "tick",
      yAxisMax: 100,
    };
  } else {
    chartOptions.tooltipOptions = {
      formatTooltipY: (d) => frappe.format(d, { fieldtype: "Currency" }),
    };
  }

  // Special options for stacked bar chart
  if (chartType === "Stacked") {
    chartOptions.stacked = 1;
  }

  // Render the chart
  frappe.query_report.render_chart({
    type: chartTypeForRender,
    title: chartTitle,
    data: chartData,
    ...chartOptions,
  });

  frappe.msgprint(
    __(
      `Now showing resource utilization as ${
        viewType === "percentage" ? "percentages" : "absolute values"
      }`
    )
  );
}

// Show monthly trend of resource utilization
function showMonthlyTrend(report) {
  if (!report.data || !report.data.length) {
    frappe.msgprint(__("No data available to display"));
    return;
  }

  // Force group by month if not already
  if (frappe.query_report.get_filter_value("group_by") !== "Month") {
    frappe.query_report.set_filter_value("group_by", "Month");
    frappe.query_report.refresh();
    return;
  }

  let data = report.data;
  let labels = [];
  let totalValues = [];
  let laborRatios = [];
  let materialRatios = [];
  let equipmentRatios = [];

  // Extract data for the trend chart
  data.forEach((d) => {
    labels.push(d.month || d.period);
    totalValues.push(d.total_cost || 0);
    laborRatios.push(d.labor_percentage || 0);
    materialRatios.push(d.material_percentage || 0);
    equipmentRatios.push(d.equipment_percentage || 0);
  });

  // Create chart data
  let chartData = {
    labels: labels,
    datasets: [
      {
        name: "Total Cost",
        values: totalValues,
        chartType: "bar",
      },
      {
        name: "Labor %",
        values: laborRatios,
        chartType: "line",
      },
      {
        name: "Material %",
        values: materialRatios,
        chartType: "line",
      },
      {
        name: "Equipment %",
        values: equipmentRatios,
        chartType: "line",
      },
    ],
  };

  // Render the chart
  frappe.query_report.render_chart({
    type: "axis-mixed",
    title: "Monthly Resource Utilization Trend",
    data: chartData,
    height: 300,
    colors: ["#FF5858", "#5EC962", "#7B6CFF", "#FFA00A"],
    axisOptions: {
      xIsSeries: 1,
    },
    tooltipOptions: {
      formatTooltipY: (d, i, dataset) => {
        if (dataset.name === "Total Cost") {
          return frappe.format(d, { fieldtype: "Currency" });
        } else {
          return d.toFixed(1) + "%";
        }
      },
    },
  });

  frappe.msgprint(__("Showing monthly trend of resource utilization"));
}

// Show resource utilization summary
function showResourceSummary(report) {
  if (!report.data || !report.data.length) {
    frappe.msgprint(__("No data available for summary"));
    return;
  }

  let data = report.data;

  // Calculate totals
  let totalCost = 0;
  let totalLabor = 0;
  let totalMaterial = 0;
  let totalEquipment = 0;
  let totalSupervision = 0;

  data.forEach((d) => {
    totalCost += d.total_cost || 0;
    totalLabor += d.labor_cost || 0;
    totalMaterial += d.material_cost || 0;
    totalEquipment += d.equipment_cost || 0;
    totalSupervision += d.supervision_cost || 0;
  });

  // Calculate percentages
  const laborPercentage = ((totalLabor / totalCost) * 100).toFixed(2);
  const materialPercentage = ((totalMaterial / totalCost) * 100).toFixed(2);
  const equipmentPercentage = ((totalEquipment / totalCost) * 100).toFixed(2);
  const supervisionPercentage = ((totalSupervision / totalCost) * 100).toFixed(
    2
  );

  // Find the highest resource utilization
  const resources = [
    { name: "Labor", cost: totalLabor, percentage: laborPercentage },
    { name: "Material", cost: totalMaterial, percentage: materialPercentage },
    {
      name: "Equipment",
      cost: totalEquipment,
      percentage: equipmentPercentage,
    },
    {
      name: "Supervision",
      cost: totalSupervision,
      percentage: supervisionPercentage,
    },
  ];

  const highestResource = resources.reduce((prev, current) =>
    prev.cost > current.cost ? prev : current
  );

  // Create a message with an embedded chart
  const message = `
	  <div style="text-align: center; padding: 15px;">
		<h3>Resource Utilization Summary</h3>
		
		<div style="margin: 20px 0;">
		  <div id="resource_summary_chart" style="height: 200px;"></div>
		</div>
		
		<div style="display: flex; justify-content: space-between; margin-bottom: 10px;">
		  <div style="text-align: left; width: 50%;">Total Cost:</div>
		  <div style="text-align: right; width: 50%;">${frappe.format(totalCost, {
        fieldtype: "Currency",
      })}</div>
		</div>
		
		<div style="display: flex; justify-content: space-between; margin-bottom: 10px;">
		  <div style="text-align: left; width: 50%;">Labor Cost (${laborPercentage}%):</div>
		  <div style="text-align: right; width: 50%;">${frappe.format(totalLabor, {
        fieldtype: "Currency",
      })}</div>
		</div>
		
		<div style="display: flex; justify-content: space-between; margin-bottom: 10px;">
		  <div style="text-align: left; width: 50%;">Material Cost (${materialPercentage}%):</div>
		  <div style="text-align: right; width: 50%;">${frappe.format(totalMaterial, {
        fieldtype: "Currency",
      })}</div>
		</div>
		
		<div style="display: flex; justify-content: space-between; margin-bottom: 10px;">
		  <div style="text-align: left; width: 50%;">Equipment Cost (${equipmentPercentage}%):</div>
		  <div style="text-align: right; width: 50%;">${frappe.format(totalEquipment, {
        fieldtype: "Currency",
      })}</div>
		</div>
		
		<div style="display: flex; justify-content: space-between; margin-bottom: 20px;">
		  <div style="text-align: left; width: 50%;">Supervision Cost (${supervisionPercentage}%):</div>
		  <div style="text-align: right; width: 50%;">${frappe.format(
        totalSupervision,
        { fieldtype: "Currency" }
      )}</div>
		</div>
		
		<div style="background-color: #f8f8f8; padding: 10px; border-radius: 5px;">
		  <strong>Key Insight:</strong> ${
        highestResource.name
      } represents the highest cost category at ${
    highestResource.percentage
  }% of total costs.
		</div>
	  </div>
	`;

  frappe.msgprint({
    title: __("Resource Utilization Summary"),
    indicator: "blue",
    message: message,
    wide: true,
  });

  // Render a donut chart inside the message
  setTimeout(() => {
    if (document.getElementById("resource_summary_chart")) {
      new frappe.Chart("#resource_summary_chart", {
        title: "Resource Distribution",
        data: {
          labels: ["Labor", "Material", "Equipment", "Supervision"],
          datasets: [
            {
              values: [
                parseFloat(laborPercentage),
                parseFloat(materialPercentage),
                parseFloat(equipmentPercentage),
                parseFloat(supervisionPercentage),
              ],
            },
          ],
        },
        type: "donut",
        colors: ["#5EC962", "#7B6CFF", "#FFA00A", "#00BCD4"],
        tooltipOptions: {
          formatTooltipY: (d) => d.toFixed(1) + "%",
        },
      });
    }
  }, 100);
}

// Function to export chart as PNG (placeholder - actual implementation would depend on Frappe's capabilities)
function exportChart(report) {
  frappe.msgprint(
    __(
      "Chart export feature will download the current chart view as a PNG file."
    )
  );

  // In a real implementation, this would use Frappe's chart API to export the chart
  setTimeout(() => {
    // This is just a placeholder for the actual export functionality
    frappe.msgprint(__("Chart exported successfully!"));
  }, 1000);
}
